inc/osvr/Client/SkeletonConfig.h

```diff
@@ -0,0 +1,271 @@
+        friend class SkeletonConfigFactory;
+        SkeletonConfig(OSVR_ClientContext ctx);
+        /// @brief check if the articulation tree has been updated, then we need
+        /// to traverse the articulation tree again, and update values
+        bool isSkeletonTreeUpdated() const;
+        OSVR_ClientContext m_ctx;
+        osvr::common::PathTree m_articulationTree;
+        osvr::common::RegisteredStringMap m_jointMap;
+        osvr::common::RegisteredStringMap m_boneMap;
+        InterfaceMap m_jointInterfaces;
+        InterfaceMap m_boneInterfaces;
+        PoseMap m_jointPoses;
+        PoseMap m_bonePoses;
+    };
+
+    inline bool SkeletonConfig::getBoneId(const char *boneName,
+                                          OSVR_SkeletonBoneCount *boneId) {
+
+        osvr::util::StringID id = m_boneMap.getStringID(boneName);
+        if (id.empty()) {
+            return false;
+        } else {
+            *boneId = id.value();
+            return true;
+        }
+    }
+
+    inline bool
+    SkeletonConfig::getAvailableJointId(OSVR_SkeletonJointCount jointIndex,
+                                        OSVR_SkeletonJointCount *jointId) {
+        *jointId = 0;
+        if (jointIndex >= m_jointInterfaces.size()) {
+            return false;
+        }
+
+        // m_jointInterfaces[jointIndex].first.value() is the return value,
+        // but we need to make sure we actually allocate an entry for it
+
+        // get the string id
+        auto stringId =
+            m_jointMap.getStringFromId(m_jointInterfaces[jointIndex].first);
+
+        // then get the joint id
+        auto ret = getJointId(stringId.c_str(), jointId);
+
+        return ret;
+    }
+
+    inline bool
+    SkeletonConfig::getAvailableBoneId(OSVR_SkeletonBoneCount boneIndex,
+                                       OSVR_SkeletonBoneCount *boneId) {
+        *boneId = 0;
+
+        if (boneIndex >= m_boneMap.getEntries().size()) {
+            return false;
+        }
+
+        auto boneName = m_boneMap.getEntries().at(boneIndex);
+
+        auto ret = getBoneId(boneName.c_str(), boneId);
+
+        return ret;
+    }
+
+    inline bool SkeletonConfig::getJointId(const char *jointName,
+                                           OSVR_SkeletonJointCount *jointId) {
+
+        osvr::util::StringID id = m_jointMap.getStringID(jointName);
+        if (id.empty()) {
+            return false;
+        } else {
+            *jointId = id.value();
+            return true;
+        }
+    }
+
+    inline OSVR_Pose3
+    SkeletonConfig::getJointState(OSVR_SkeletonJointCount jointId) {
+
+        OSVR_Pose3 pose;
+        osvrPose3SetIdentity(&pose);
+
+        // find an interface for given jointId
+        for (auto val : m_jointPoses) {
+            if (val.first.value() == jointId) {
+                // return the joint state
+                return val.second;
+            }
+        }
+        // pose not available for this frame
+        throw NoPoseYet();
```

Comment:
If the jointId/boneId is provided, then there should be a pose associated with it. All joint poses are reported together in a single frame, so lack of pose indicates something is not right.